# **Deep Learning - CSE4006**
# LAB 2

> ## Name: **K CHARVI**
> ## Reg No: 19BCE7002
> ## Date: 22|10|2001
> ## Submitted To: Dr. BKSP Kumarraju Alluri


---



---



In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### Imports

In [2]:
import numpy as np
import pandas as pd
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import train_test_split
from keras.layers import Dropout
from tensorflow.keras.optimizers import SGD

# Question 1: Use the above code of "KerasClassifier" on new dataset

### Loading Dataset

In [3]:
dataset=pd.read_csv('/content/drive/MyDrive/Deep learning LAB/L2D2.csv',dtype='object', header=None)
dataset

0       1       2       3       4   ...      56      57      58      59 60
0    0.0200  0.0371  0.0428  0.0207  0.0954  ...  0.0180  0.0084  0.0090  0.0032  R
1    0.0453  0.0523  0.0843  0.0689  0.1183  ...  0.0140  0.0049  0.0052  0.0044  R
2    0.0262  0.0582  0.1099  0.1083  0.0974  ...  0.0316  0.0164  0.0095  0.0078  R
3    0.0100  0.0171  0.0623  0.0205  0.0205  ...  0.0050  0.0044  0.0040  0.0117  R
4    0.0762  0.0666  0.0481  0.0394  0.0590  ...  0.0072  0.0048  0.0107  0.0094  R
..      ...     ...     ...     ...     ...  ...     ...     ...     ...     ... ..
203  0.0187  0.0346  0.0168  0.0177  0.0393  ...  0.0065  0.0115  0.0193  0.0157  M
204  0.0323  0.0101  0.0298  0.0564  0.0760  ...  0.0034  0.0032  0.0062  0.0067  M
205  0.0522  0.0437  0.0180  0.0292  0.0351  ...  0.0140  0.0138  0.0077  0.0031  M
206  0.0303  0.0353  0.0490  0.0608  0.0167  ...  0.0034  0.0079  0.0036  0.0048  M
207  0.0260  0.0363  0.0136  0.0272  0.0214  ...  0.0040  0.0036  0.0061  0.0115  M

[208 rows x 61 columns]

In [4]:
dataset.isnull().any().any()

False

In [5]:
dataset.replace(['R','M'],['0','1'],inplace=True)
dataset.head()

0       1       2       3       4   ...      56      57      58      59 60
0  0.0200  0.0371  0.0428  0.0207  0.0954  ...  0.0180  0.0084  0.0090  0.0032  0
1  0.0453  0.0523  0.0843  0.0689  0.1183  ...  0.0140  0.0049  0.0052  0.0044  0
2  0.0262  0.0582  0.1099  0.1083  0.0974  ...  0.0316  0.0164  0.0095  0.0078  0
3  0.0100  0.0171  0.0623  0.0205  0.0205  ...  0.0050  0.0044  0.0040  0.0117  0
4  0.0762  0.0666  0.0481  0.0394  0.0590  ...  0.0072  0.0048  0.0107  0.0094  0

[5 rows x 61 columns]

In [6]:
dataset=dataset.apply(pd.to_numeric,errors='coerce').fillna(0)

In [7]:
dataset.dtypes

0     float64
1     float64
2     float64
3     float64
4     float64
       ...   
56    float64
57    float64
58    float64
59    float64
60      int64
Length: 61, dtype: object

In [8]:
X = dataset.iloc[:, 0:60]
Y = dataset.iloc[:, 60]
print(X)
print(Y)

         0       1       2       3   ...      56      57      58      59
0    0.0200  0.0371  0.0428  0.0207  ...  0.0180  0.0084  0.0090  0.0032
1    0.0453  0.0523  0.0843  0.0689  ...  0.0140  0.0049  0.0052  0.0044
2    0.0262  0.0582  0.1099  0.1083  ...  0.0316  0.0164  0.0095  0.0078
3    0.0100  0.0171  0.0623  0.0205  ...  0.0050  0.0044  0.0040  0.0117
4    0.0762  0.0666  0.0481  0.0394  ...  0.0072  0.0048  0.0107  0.0094
..      ...     ...     ...     ...  ...     ...     ...     ...     ...
203  0.0187  0.0346  0.0168  0.0177  ...  0.0065  0.0115  0.0193  0.0157
204  0.0323  0.0101  0.0298  0.0564  ...  0.0034  0.0032  0.0062  0.0067
205  0.0522  0.0437  0.0180  0.0292  ...  0.0140  0.0138  0.0077  0.0031
206  0.0303  0.0353  0.0490  0.0608  ...  0.0034  0.0079  0.0036  0.0048
207  0.0260  0.0363  0.0136  0.0272  ...  0.0040  0.0036  0.0061  0.0115

[208 rows x 60 columns]
0      0
1      0
2      0
3      0
4      0
      ..
203    1
204    1
205    1
206    1
207    1


### Split to training and testing

In [9]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.33, random_state=7)

## MODEL 1

In [10]:
model1 = Sequential()
model1.add(Dense(64, input_dim=60, kernel_initializer='uniform', activation='relu'))
model1.add(Dense(60, kernel_initializer='uniform', activation='relu'))
model1.add(Dense(1, kernel_initializer='uniform', activation='sigmoid'))

In [11]:
model1.compile(
          loss='binary_crossentropy',
          optimizer='adam',
          metrics=['accuracy']
)

In [12]:
model1.fit(X_train,
          Y_train,
          epochs=150,
          batch_size=10)

Epoch 1/150
14/14 [==============================] - 0s 2ms/step - loss: 0.6930 - accuracy: 0.4820
Epoch 2/150
14/14 [==============================] - 0s 1ms/step - loss: 0.6915 - accuracy: 0.5540
Epoch 3/150
14/14 [==============================] - 0s 2ms/step - loss: 0.6882 - accuracy: 0.5468
Epoch 4/150
14/14 [==============================] - 0s 1ms/step - loss: 0.6771 - accuracy: 0.6331
Epoch 5/150
14/14 [==============================] - 0s 1ms/step - loss: 0.6546 - accuracy: 0.6547
Epoch 6/150
14/14 [==============================] - 0s 1ms/step - loss: 0.6221 - accuracy: 0.7122
Epoch 7/150
14/14 [==============================] - 0s 1ms/step - loss: 0.5862 - accuracy: 0.7194
Epoch 8/150
14/14 [==============================] - 0s 2ms/step - loss: 0.5534 - accuracy: 0.7770
Epoch 9/150
14/14 [==============================] - 0s 2ms/step - loss: 0.5193 - accuracy: 0.7698
Epoch 10/150
14/14 [==============================] - 0s 2ms/step - loss: 0.4822 - accuracy: 0.7698
Epoch 11/

In [13]:
scores = model1.evaluate(X_test, Y_test)
print("%s: %.2f%%" % (model1.metrics_names[1], scores[1]*100))

3/3 [==============================] - 0s 3ms/step - loss: 0.6612 - accuracy: 0.8551
accuracy: 85.51%


### **The accuracy obtained is 85.51% from a model with three layers**


# Question 2: Check whether accuracy will improve with the larger network

## MODEL 2


In [14]:
model2 = Sequential()
model2.add(Dense(256, input_dim=60, kernel_initializer='random_uniform', activation='relu'))
model2.add(Dense(192, kernel_initializer='uniform', activation='relu'))
model2.add(Dense(128, kernel_initializer='uniform', activation='relu'))
model2.add(Dense(92, kernel_initializer='uniform', activation='relu'))
model2.add(Dense(60, kernel_initializer='uniform', activation='relu'))
model2.add(Dense(172, kernel_initializer='uniform', activation='relu'))
model2.add(Dense(148, kernel_initializer='uniform', activation='relu'))
model2.add(Dense(84, kernel_initializer='uniform', activation='relu'))
model2.add(Dense(48, kernel_initializer='uniform', activation='relu'))
model2.add(Dense(1, kernel_initializer='uniform', activation='sigmoid'))

model2.compile(
          loss='binary_crossentropy',
          optimizer='adam',
          metrics=['accuracy']
)

model2.summary()


Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 256)               15616     
_________________________________________________________________
dense_4 (Dense)              (None, 192)               49344     
_________________________________________________________________
dense_5 (Dense)              (None, 128)               24704     
_________________________________________________________________
dense_6 (Dense)              (None, 92)                11868     
_________________________________________________________________
dense_7 (Dense)              (None, 60)                5580      
_________________________________________________________________
dense_8 (Dense)              (None, 172)               10492     
_________________________________________________________________
dense_9 (Dense)              (None, 148)              

In [15]:
model2.fit(X_train,
          Y_train,
          epochs=150,
          batch_size=10)

Epoch 1/150
14/14 [==============================] - 1s 4ms/step - loss: 0.6932 - accuracy: 0.4820
Epoch 2/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6930 - accuracy: 0.5252
Epoch 3/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6929 - accuracy: 0.5252
Epoch 4/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6929 - accuracy: 0.5252
Epoch 5/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6928 - accuracy: 0.5252
Epoch 6/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6927 - accuracy: 0.5252
Epoch 7/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6928 - accuracy: 0.5252
Epoch 8/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6925 - accuracy: 0.5252
Epoch 9/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6924 - accuracy: 0.5252
Epoch 10/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6902 - accuracy: 0.5252
Epoch 11/

In [16]:
scores = model2.evaluate(X_test, Y_test)
print("%s: %.2f%%" % (model2.metrics_names[1], scores[1]*100))

3/3 [==============================] - 0s 4ms/step - loss: 1.5591 - accuracy: 0.8406
accuracy: 84.06%


### **The accuracy obtained is 84.06% from a larger network, which is a decrease from the previous.**

> **NOTE: We cannot always confirm that it would decrease in some cases taking different layers the accuracy had increased.**






# Question 3: Check whether accuracy will be improved after Dropout or not

## MODEL 3


In [17]:
model3 = Sequential()

#DROP OUT LAYER
model3.add(Dropout(0.2, input_shape=(60,)))

model3.add(Dense(256, kernel_initializer='random_uniform', activation='relu'))
model3.add(Dense(192, kernel_initializer='uniform', activation='relu'))
model3.add(Dense(128, kernel_initializer='uniform', activation='relu'))
model3.add(Dense(92, kernel_initializer='uniform', activation='relu'))
model3.add(Dense(60, kernel_initializer='uniform', activation='relu'))
model3.add(Dense(172, kernel_initializer='uniform', activation='relu'))
model3.add(Dense(148, kernel_initializer='uniform', activation='relu'))
model3.add(Dense(84, kernel_initializer='uniform', activation='relu'))
model3.add(Dense(48, kernel_initializer='uniform', activation='relu'))
model3.add(Dense(1, kernel_initializer='uniform', activation='sigmoid'))

model3.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])


In [18]:
model3.fit(X_train,
          Y_train,
          epochs=150,
          batch_size=10)

Epoch 1/150
14/14 [==============================] - 1s 3ms/step - loss: 0.6934 - accuracy: 0.4245
Epoch 2/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6931 - accuracy: 0.5252
Epoch 3/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6931 - accuracy: 0.5252
Epoch 4/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6930 - accuracy: 0.5252
Epoch 5/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6930 - accuracy: 0.5252
Epoch 6/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6929 - accuracy: 0.5252
Epoch 7/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6929 - accuracy: 0.5252
Epoch 8/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6928 - accuracy: 0.5252
Epoch 9/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6928 - accuracy: 0.5252
Epoch 10/150
14/14 [==============================] - 0s 3ms/step - loss: 0.6928 - accuracy: 0.5252
Epoch 11/

In [19]:
scores = model3.evaluate(X_test, Y_test)
print("%s: %.2f%%" % (model3.metrics_names[1], scores[1]*100))

3/3 [==============================] - 0s 4ms/step - loss: 0.3236 - accuracy: 0.8261
accuracy: 82.61%


**Accuracy is almost similar to the previous network though not an effective network.**

# Question 4: Check whether the accuracy improved or not with decay learning rate; Use SGD optimizer

#### Decay = Learning Rate/ Number of epochs

## MODEL 4

In [20]:


model4 = Sequential()
model4.add(Dropout(0.2, input_shape=(60,)))
model4.add(Dense(256, kernel_initializer='random_uniform', activation='relu'))
model4.add(Dense(192, kernel_initializer='uniform', activation='relu'))
model4.add(Dense(128, kernel_initializer='uniform', activation='relu'))
model4.add(Dense(92, kernel_initializer='uniform', activation='relu'))
model4.add(Dense(60, kernel_initializer='uniform', activation='relu'))
model4.add(Dense(172, kernel_initializer='uniform', activation='relu'))
model4.add(Dense(148, kernel_initializer='uniform', activation='relu'))
model4.add(Dense(84, kernel_initializer='uniform', activation='relu'))
model4.add(Dense(48, kernel_initializer='uniform', activation='relu'))
model4.add(Dense(1, kernel_initializer='uniform', activation='sigmoid'))

#APPLYING DECAY LR
sgd = SGD(learning_rate=0.1, momentum=0.9, decay=0.1/150)
model4.compile(loss='binary_crossentropy', optimizer='sgd', metrics=['accuracy'])


In [21]:
scores = model4.evaluate(X_test, Y_test)
print("%s: %.2f%%" % (model4.metrics_names[1], scores[1]*100))

3/3 [==============================] - 0s 3ms/step - loss: 0.6931 - accuracy: 0.4638
accuracy: 46.38%


**Accuracy has decreased drastically**



---

